src/tools/rbd_wnbd/rbd_wnbd.cc

```diff
@@ -902,6 +902,37 @@ class RBDService : public ServiceBase {
     }
 };
 
+class WNBDWatchCtx : public librbd::UpdateWatchCtx
+{
+private:
+  librados::IoCtx &io_ctx;
+  WnbdHandler* handler;
+  librbd::Image &image;
+  uint64_t size;
+public:
+  WNBDWatchCtx(librados::IoCtx& io_ctx, WnbdHandler* handler,
+               librbd::Image& image, uint64_t size)
+    : io_ctx(io_ctx)
+    , handler(handler)
+    , image(image)
+    , size(size)
+  { }
+
+  ~WNBDWatchCtx() override {}
+
+  void handle_notify() override
+  {
+    uint64_t new_size;
+
+    if (image.size(&new_size) == 0) {
+      if (new_size != size) {
+        if (!handler->resize(new_size))
+          size = new_size;
+      }
+    }
```

Comment:
Nit: I'd combine this into a single `if` statement:
```suggestion
    if (image.size(&new_size) == 0 && new_size != size &&
        handler->resize(new_size) == 0) {
      size = new_size;
    }
```